// @nx/gradle/src/executors/gradle/gradle-batch.impl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.batchRunnerPath = void 0;
exports.default = gradleBatch;
const devkit_1 = require("@nx/devkit");
const run_commands_impl_1 = require("nx/src/executors/run-commands/run-commands.impl");
const exec_gradle_1 = require("../../utils/exec-gradle");
const path_1 = require("path");
const child_process_1 = require("child_process");
const pseudo_terminal_1 = require("nx/src/tasks-runner/pseudo-terminal");
exports.batchRunnerPath = (0, path_1.join)(__dirname, '../../../batch-runner/build/libs/batch-runner-all.jar');
async function gradleBatch(taskGraph, inputs, overrides, context) {
    try {
        const projectName = taskGraph.tasks[taskGraph.roots[0]]?.target?.project;
        let projectRoot = context.projectGraph.nodes[projectName]?.data?.root ?? '';
        const gradlewPath = (0, exec_gradle_1.findGradlewFile)((0, path_1.join)(projectRoot, 'project.json')); // find gradlew near project root
        const root = (0, path_1.join)(context.root, (0, path_1.dirname)(gradlewPath));
        // set args with passed in args and overrides in command line
        const input = inputs[taskGraph.roots[0]];
        let args = typeof input.args === 'string'
            ? input.args.trim().split(' ')
            : Array.isArray(input.args)
                ? input.args
                : [];
        if (overrides.__overrides_unparsed__.length) {
            args.push(...overrides.__overrides_unparsed__);
        }
        const gradlewTasksToRun = Object.entries(taskGraph.tasks).reduce((gradlewTasksToRun, [taskId, task]) => {
            const gradlewTaskName = inputs[task.id].taskName;
            const testClassName = inputs[task.id].testClassName;
            gradlewTasksToRun[taskId] = {
                taskName: gradlewTaskName,
                testClassName: testClassName,
            };
            return gradlewTasksToRun;
        }, {});
        const gradlewBatchStart = performance.mark(`gradlew-batch:start`);
        const usePseudoTerminal = process.env.NX_NATIVE_COMMAND_RUNNER !== 'false' &&
            pseudo_terminal_1.PseudoTerminal.isSupported();
        const command = `java -jar ${exports.batchRunnerPath} --tasks='${JSON.stringify(gradlewTasksToRun)}' --workspaceRoot=${root} --args='${args
            .join(' ')
            .replaceAll("'", '"')}' ${process.env.NX_VERBOSE_LOGGING === 'true' ? '' : '--quiet'}`;
        let batchResults;
        if (usePseudoTerminal) {
            const terminal = (0, pseudo_terminal_1.createPseudoTerminal)();
            await terminal.init();
            const cp = terminal.runCommand(command, {
                cwd: devkit_1.workspaceRoot,
                jsEnv: process.env,
                quiet: process.env.NX_VERBOSE_LOGGING !== 'true',
            });
            const results = await cp.getResults();
            batchResults = results.terminalOutput;
            batchResults = batchResults.replace(command, '');
            const startIndex = batchResults.indexOf('{');
            const endIndex = batchResults.lastIndexOf('}');
            batchResults = batchResults.substring(startIndex, endIndex + 1);
        }
        else {
            batchResults = (0, child_process_1.execSync)(command, {
                cwd: devkit_1.workspaceRoot,
                windowsHide: true,
                env: process.env,
                maxBuffer: run_commands_impl_1.LARGE_BUFFER,
            }).toString();
        }
        const gradlewBatchEnd = performance.mark(`gradlew-batch:end`);
        performance.measure(`gradlew-batch`, gradlewBatchStart.name, gradlewBatchEnd.name);
        const gradlewBatchResults = JSON.parse(batchResults.toString());
        Object.keys(taskGraph.tasks).forEach((taskId) => {
            if (!gradlewBatchResults[taskId]) {
                gradlewBatchResults[taskId] = {
                    success: false,
                    terminalOutput: `Gradlew batch failed`,
                };
            }
        });
        return gradlewBatchResults;
    }
    catch (e) {
        devkit_1.output.error({
            title: `Gradlew batch failed`,
            bodyLines: [e.toString()],
        });
        return taskGraph.roots.reduce((acc, key) => {
            acc[key] = { success: false, terminalOutput: e.toString() };
            return acc;
        }, {});
    }
}
